### ISSUE TITLE
[New build - EXPERIMENTAL] 20w40b: Turrets, Rotators and Hotfixes

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#3024]
A much needed hotfixes build, and some popular turret and rotator improvements.

Turrets Improvements:
 - Turrets can now be build on children entity. 
 - Retracting mechanism are now supported, the turret will recompute its collision map once the children entity has finished moving.
 - You can define a resting position inside the turret computer menu.
 
 Rotator Improvements:
 - Can setup current angle in the rotator menu.
 - Can setup angular velocity.
 - Can setup resting position for turret use.

Bugfixes:
#2991 Can place blocks on small size mechanism.
#3007 Galaxy still visible on holo projector delete.
#3010 Build placement widget still visible on item switch.
#3008 Cursor reset to center on Holomap exit.
#3011 Monitor linked to an Turret Computer cannot be activated.
#2977 Turrets on children entity improper collisions.

**Comment by tsunamayo** [comment]
Build is live! Thanks

**Comment by CurioInventorium** [comment]
I was going to put a ship on the steam workshop tomorrow that consists of stable branch blocks, but built on the experimental branch.
Will people on stable branch be able to use it okay once I've updated to this release?
I know when I was writing my dissertation in Unity, that newer serialized values would be ignored on loading with an earlier version.

**Comment by tsunamayo** [comment]
@CurioInventorium I it should be fine, but if you used some of the new feature then of course people would loose that.

**Comment by CurioInventorium** [comment]
@tsunamayo Cool, thank you.

**Comment by Vicomt** [comment]
lookin good. did you have chance to look at the blueprint I sent in issue #3013 ? it would be nice to know whether that blueprint is broken and we need to revert to an older version.

**Comment by BrickMacklin** [comment]
> Turrets can now be build on children entity.

What do you mean by this? We were already doing that.

**Comment by Kaiser-Indrasil** [comment]
> Turrets can now be build on children entity.

That's not new. I was able to build turrets on child entities before perfectly fine.
I've tried to do something like this below, but it looks like you didn't mean that either.
So, what's new exactly?
![image](https://user-images.githubusercontent.com/63756057/95005452-d9d77c80-05f0-11eb-9a80-6757ae53a8d7.png)
Also, what do you mean by this?:
> Retracting mechanism are now supported
Since we were already able to place turrets on child entities, they could also be retractable.
I thought you meant that we can now link sequencer gates to mechanisms that are in use of a turret computer but that doesn't seem to be the case either:
![image](https://user-images.githubusercontent.com/63756057/95005570-3be4b180-05f2-11eb-8680-ae8987494c47.png)

**Comment by tsunamayo** [comment]
@BrickMacklin @Kaiser-Indrasil it was never really supported, sometime it worked sometime not. I guess you were on the lucky side ;) I had several issues reported with that.
@Kaiser-Indrasil nope this still wont work.
 By retracting I mean something where the turret pop out of the ship (I saw a few build like that).
Could you give me a test case of the current angle not working (it worked for me).
Thanks

**Comment by Kaiser-Indrasil** [comment]
Yeah, this is what I had in mind thinking about retractable turrets.
I was able to do it all along but I think you mean that this feature is now officially supported, yes?

I would give you the test case blueprint but it seems that it updates correctly on paste-in.
It doesn't work after I type in the number though.

**Comment by Kaiser-Indrasil** [comment]
Oh, sorry, rotators adjust their position just fine - it's the hinges I've been having problem with.

**Comment by tsunamayo** [comment]
ah okay I will check the hinge!

**Comment by Kaiser-Indrasil** [comment]
Also, there seems to be something wrong with resting position. It renders the automatic turret useless since it barrs it from tracking the target and rotates it to weird degrees.
![image](https://user-images.githubusercontent.com/63756057/95006438-6e93a780-05fc-11eb-9948-0a43c5368019.png)

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil not sure what is happening? Resting position is for when the turret is no longer used, it will come back to a basic position.

**Comment by matejkar** [comment]
@tsunamayo the turret update is nice, especially the option to default turret resting position, but when can we see updated targeting for turrets (since as i said many times, now turrets only attack aggresive npcs and its not really possible to have large turret based battles)?

**Comment by francklaballe** [comment]
Before this update my turret was working perfectly.

From my cockpit seat and yoke, I could point my lasers exactly where I wanted.
Now some axes when they reach certain degrees the rotors no longer work

for information I had my turret computer outside the turret so outside the child area.

I'll try to put the turret computer built into the turret to see.

**Comment by Kaiser-Indrasil** [comment]
Well, the turret is now shoved into the hardpoint and I have no way of unclipping it out. I can give you the test case if you want.

**Comment by matejkar** [comment]
@Kaiser-Indrasil how would you give a test case?

**Comment by Kaiser-Indrasil** [comment]
So this here video explains my issue with this update. Rotors only start working past 360 degrees and they always start rotating from 0 degrees.
Hinges only start working from 180 degrees (!) up.

[18QKwTxadH.zip](https://github.com/tsunamayo/Starship-EVO/files/5323788/18QKwTxadH.zip)

**Comment by tsunamayo** [comment]
@francklaballe Hi, could you please send me the blueprint! The place where the computer is has no impact.

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil I dont have any issue with turret on my side, could you please send me your stuck blueprint? Thanks

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil thanks for the vid btw, I am looking at that atm

**Comment by Kaiser-Indrasil** [comment]
Here you go. The Turret Computers are at the base of the turret (long ones) and at the top back (bulky short ones).
[Turrets 40C.zip](https://github.com/tsunamayo/Starship-EVO/files/5324644/Turrets.40C.zip)

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil thanks! issue with the rotator / hinge new input angle is fixed. For the turret they works fine on your ship when I tested! Not sure what is happening on your side? If you could make me a video ;) thanks

**Comment by Kaiser-Indrasil** [comment]
Sure, here you go. This time, the video was too large so I just uploaded it to YouTube instead ;)
https://youtu.be/QIJVeGEEHX8
As you can see, the turrets don't work at all despite having been set to autofire upon an aggressive target.
Nor can I reliably adjust the angles of both the rotator and the hinge.
The turrets were able to work perfectly fine before the 40c version.

**Comment by tsunamayo** [comment]
Hum I can use your ship perfectly, I dont really get it... Maybe I it is one of the fix I did? I will  push the new build asap and you will tell me if it work better.

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil  could you tell me if it works in 20w41a? Thanks
